crates/biome_graphql_parser/src/parser/type.rs

```diff
@@ -0,0 +1,53 @@
+use crate::parser::{parse_name, GraphqlParser};
+use biome_graphql_syntax::{GraphqlSyntaxKind::*, T};
+use biome_parser::{parsed_syntax::ParsedSyntax, prelude::ParsedSyntax::*, Parser};
+
+use super::{
+    is_at_name,
+    parse_error::{expected_named_or_list_type, expected_type},
+};
+
+#[inline]
+pub(crate) fn parse_type(p: &mut GraphqlParser) -> ParsedSyntax {
+    let m = p.start();
+    let node = if is_at_list_type(p) {
+        parse_list_type(p)
+    } else {
+        parse_named_type(p)
+    };
+    if p.at(T![!]) {
+        p.bump(T![!]);
+        // case like `a: !`
+        node.or_add_diagnostic(p, expected_named_or_list_type);
+        Present(m.complete(p, GRAPHQL_NON_NULL_TYPE))
+    } else {
+        m.abandon(p);
+        node
+    }
+}
+
+#[inline]
+fn parse_list_type(p: &mut GraphqlParser) -> ParsedSyntax {
+    let m = p.start();
```

Comment:
Should we add a check for `p.at(T['['])` and return `Absent`? If not, then we don't need to return a `ParsedSyntax` but return a `CompletedMarker`